File: framework-typescript/src/envelope.ts
// envelope.ts — the CONTRACT ENVELOPE the served tool returns, and the three
// refusals plus the one normalization that stand between a walk's Result and it.
//
// NO KEY IS EVER OMITTED FROM THE THREE TOP-LEVEL FIELDS, and walk_complete is
// the one that matters. The contract output schema REQUIRES walk_complete, and
// this package validates the envelope against that schema before the result
// leaves the provider; a walk asserting INCOMPLETE that emitted no walk_complete
// at all would fail the whole collect with a missing-property error. A walk
// asserting COMPLETE emits `true` and the defect is invisible, which is why the
// test that pins this rides the incomplete arm.
//
// EVERY LIST IS AN ARRAY, NEVER null AND NEVER AN OMITTED KEY. A walk that found
// nothing leaves both properties undefined; `undefined` disappears from
// JSON.stringify and `null` marshals as null; the contract output schema declares
// both as "array" with no null union. An empty region, an empty log group or a
// filter that matched nothing is the first real run of a new collector, so this is
// the cell it hits. Nothing upstream catches it: measured, a result carrying
// `edges: null` went out untouched on every route tried.

import { isCompleteness } from "./completeness.js";
import type { CollectorEdge, CollectorNode, Result } from "./framework.js";

/** The served tool's output value: the contract envelope, exactly. */
export interface CollectOutput {
  nodes: Record<string, unknown>[];
  edges: Record<string, unknown>[];
  walk_complete: boolean;
}

/**
 * The wire key each node property is emitted under. The map IS the contract's
 * node vocabulary; a property this package does not name never reaches the wire,
 * and a key the contract does not declare is never emitted.
 */
export const NODE_WIRE_KEYS: ReadonlyMap<keyof CollectorNode, string> = new Map([
  ["id", "id"],
  ["type", "type"],
  ["symbolName", "symbol_name"],
  ["filePath", "file_path"],
  ["language", "language"],
  ["startLine", "start_line"],
  ["endLine", "end_line"],
  ["content", "content"],
  ["signature", "signature"],
  ["summary", "summary"],
  ["description", "description"],
  ["source", "source"],
  ["status", "status"],
  ["keywords", "keywords"],
  ["isExported", "is_exported"],
  ["metadata", "metadata"],
]);

/** The wire key each edge property is emitted under. */
export const EDGE_WIRE_KEYS: ReadonlyMap<keyof CollectorEdge, string> = new Map([
  ["fromId", "from_id"],
  ["toId", "to_id"],
  ["type", "type"],
  ["weight", "weight"],
  ["confidence", "confidence"],
  ["method", "method"],
  ["evidence", "evidence"],
  ["sourceGraph", "source_graph"],
  ["targetGraph", "target_graph"],
]);

/**
 * Turns one walk's Result into the envelope, refusing the three shapes that are
 * bad input and normalizing the one that is a language artifact.
 *
 * `collector` names the served tool, which is the only identity this package has
 * for the collector, so a refusal reaching an operator names something they can
 * find in their config entry.
 */
export function encodeResult(collector: string, result: Result): CollectOutput {
  const assertion = result.complete;
  if (!isCompleteness(assertion)) {
    throw new Error(
      `framework: collector "${collector}" returned a completeness value that asserts nothing; ` +
        `a walk returns complete() or incomplete(reason)`,
    );
  }
  if (!assertion.isComplete() && assertion.reason() === "") {
    throw new Error(
      `framework: collector "${collector}" asserted an INCOMPLETE walk with no reason; ` +
        `incomplete() takes the reason the walk did not finish`,
    );
  }

  const nodes: Record<string, unknown>[] = [];
  const inputNodes = result.nodes ?? [];
  for (let i = 0; i < inputNodes.length; i++) {
    const node = inputNodes[i]!;
    if (node.type === "") {
      throw new Error(
        `framework: collector "${collector}": node[${i}] (id="${node.id}") has an empty type`,
      );
    }
    nodes.push(project(node, NODE_WIRE_KEYS));
  }

  const edges: Record<string, unknown>[] = [];
  for (const edge of result.edges ?? []) {
    edges.push(project(edge, EDGE_WIRE_KEYS));
  }

  return { nodes, edges, walk_complete: assertion.isComplete() };
}

/**
 * Emits the properties the wire map names, under their contract keys, in the
 * map's own order.
 *
 * `undefined` IS THE ONLY THING DROPPED. A property set to the empty string, to
 * zero, to false or to `{}` is emitted with that value, so a node carrying its
 * optionals PRESENT AND EMPTY stays distinguishable from one that omits them —
 * which is a distinction the client's own contract tests assert. A property the
 * map does not name never reaches the wire at all, so a collector's stray key
 * cannot reach the client's strict decode and fail a collect there.
 */
function project<T extends object>(
  value: T,
  wireKeys: ReadonlyMap<keyof T, string>,
): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [property, wireKey] of wireKeys) {
    const v = value[property];
    if (v !== undefined) out[wireKey] = v;
  }
  return out;
}

File: framework-rust/src/envelope.rs
// SPDX-License-Identifier: Apache-2.0

//! envelope.rs — the CONTRACT ENVELOPE the served tool returns, and the refusals
//! that stand between a walk's [`WalkResult`] and it.
//!
//! NO SKIP-IF-EMPTY ON ANY OF THE THREE FIELDS, and `walk_complete` is the one
//! that matters. The contract output schema REQUIRES `walk_complete`; with a
//! skip attribute a walk asserting INCOMPLETE would emit no `walk_complete` at
//! all and the whole collect would fail with a missing-property error at the
//! CLIENT. A walk asserting COMPLETE emits `true` and the defect is invisible,
//! which is why the test that pins this rides the incomplete arm.
//!
//! THIS ENCODE PATH IS MORE LOAD-BEARING HERE THAN IN GO. The Go SDK's generic
//! `AddTool` validates the handler's Out value against the advertised output
//! schema before the result leaves the provider, so Go's `encodeResult` is a
//! first line with a second behind it. rmcp performs no such check: a handler
//! returning `{"nodes": {}, "edges": []}` — `walk_complete` absent, `nodes` an
//! object — reaches the wire unchanged with `isError: false`. This crate's typed
//! envelope is therefore the ONLY thing between a walk and a non-conforming
//! result, and it earns that by construction: the three fields are typed, none
//! is optional, and a `Vec` has no null spelling.
//!
//! THE EMPTY WALK NEEDS NO NORMALIZATION HERE, and the reason is worth stating
//! rather than leaving as a silence. Go normalizes a nil slice to an empty one
//! because nil marshals to `null` and the contract declares both properties as
//! `array` with no null union — an empty region or a filter that matched nothing
//! is the first real run of a new collector, so that is the cell it hits. A Rust
//! `Vec` is never null, so the rule holds for free; what would REINTRODUCE it is
//! an `Option<Vec<_>>` field or a `skip_serializing_if` attribute, and neither is
//! here. The crate's suite still asserts the emitted BYTES carry `[]`, because
//! "for free" is a property of this file's current spelling, not of Rust.

use serde::{Deserialize, Serialize};

use crate::completeness::Completeness;
use crate::error::{Error, Result};
use crate::model::{Edge, Node, WalkResult};

/// CollectOutput is the served tool's output: the contract envelope, exactly.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CollectOutput {
    pub nodes: Vec<Node>,
    pub edges: Vec<Edge>,
    pub walk_complete: bool,
}

/// encode_result turns one walk's [`WalkResult`] into the envelope, refusing the
/// shapes that are bad input.
///
/// `collector` names the served tool, which is the only identity this crate has
/// for the collector, so a refusal reaching an operator names something they can
/// find in their config entry.
///
/// THE GO FRAMEWORK HAS THREE REFUSALS HERE AND THIS ONE HAS TWO, which is a
/// difference in the type system rather than in the contract. Go's first refusal
/// is a zero `Completeness` value that asserts nothing; [`Completeness`] has no
/// such value, so that arm is absent at compile time. The other two are here.
pub fn encode_result(collector: &str, result: WalkResult) -> Result<CollectOutput> {
    if let Completeness::Incomplete { ref reason } = result.complete {
        if reason.is_empty() {
            return Err(Error::new(format!(
                "framework: collector {collector:?} asserted an INCOMPLETE walk with no reason; \
                 Completeness::incomplete takes the reason the walk did not finish"
            )));
        }
    }
    for (i, node) in result.nodes.iter().enumerate() {
        if node.node_type.is_empty() {
            return Err(Error::new(format!(
                "framework: collector {collector:?}: node[{i}] (id={:?}) has an empty type",
                node.id
            )));
        }
    }
    Ok(CollectOutput {
        walk_complete: result.complete.is_complete(),
        nodes: result.nodes,
        edges: result.edges,
    })
}
